plugins/cpp/parser/src/diagnosticmessagehandler.cpp

```diff
@@ -0,0 +1,76 @@
+#include <llvm/ADT/SmallString.h>
+#include <cppparser/filelocutil.h>
+#include "diagnosticmessagehandler.h"
+
+namespace cc
+{
+namespace parser
+{
+
+DiagnosticMessageHandler::DiagnosticMessageHandler(
+  SourceManager& srcMgr_,
+  std::shared_ptr<odb::database> db_)
+    : _srcMgr(srcMgr_), _db(db_)
+{
+}
+
+void DiagnosticMessageHandler::HandleDiagnostic(
+  clang::DiagnosticsEngine::Level diagLevel_,
+  const clang::Diagnostic& info_)
+{
+  // TODO: Is this needed?
+  clang::DiagnosticConsumer::HandleDiagnostic(diagLevel_, info_);
+
+  model::BuildLog buildLog;
+
+  //--- Message type ---//
+
+  switch (diagLevel_)
+  {
+    case clang::DiagnosticsEngine::Note:
+      buildLog.log.type = model::BuildLogMessage::Note;
+      break;
+    case clang::DiagnosticsEngine::Warning:
+      buildLog.log.type = model::BuildLogMessage::Warning;
+      break;
+    case clang::DiagnosticsEngine::Error:
+      buildLog.log.type = model::BuildLogMessage::Error;
+      break;
+    case clang::DiagnosticsEngine::Fatal:
+      buildLog.log.type = model::BuildLogMessage::FatalError;
+      break;
+  }
+
+  //--- Message content ---//
+
+  llvm::SmallString<512> content;
+  info_.FormatDiagnostic(content);
+  buildLog.log.message = content.c_str();
```

Comment:
Isn't this a lifetime violation? `content` dies but `buildLog` is saved in a data structure...

Reply:
`buildLog.log.message` is an `std::string` so a copy is made here.

Reply:
Okay... or rather, not great but not terrible...